# 2. Data wrangling

In [1]:
# import libraries 

import pandas as pd
import numpy as np

**Read data**

In [2]:
#use pd to read csv file into a dataframe df

path1 =r'C:\\Users\Domnica Maria\Documents\1. PROJECTS\Maven-Vacation-Challenge\Airbnb Data\Listings.csv'
df = pd.read_csv(path1,encoding='latin1') 

#used the encoding latin1 to read this file as there are some special character in the file

C:\Users\Domnica Maria\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (5,13) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [3]:
#view df head

df.head()

listing_id                                              name   host_id  \
0      281420    Beautiful Flat in le Village Montmartre, Paris   1466919   
1     3705183                   39 m² Paris (Sacre CÃ
“ur)  10328771   
2     4082273               Lovely apartment with Terrace, 60m2  19252768   
3     4797344               Cosy studio (close to Eiffel tower)  10668311   
4     4823489  Close to Eiffel Tower - Beautiful flat : 2 rooms  24837558   

   host_since                 host_location host_response_time  \
0  2011-12-03  Paris, Ile-de-France, France                NaN   
1  2013-11-29  Paris, Ile-de-France, France                NaN   
2  2014-07-31  Paris, Ile-de-France, France                NaN   
3  2013-12-17  Paris, Ile-de-France, France                NaN   
4  2014-12-14  Paris, Ile-de-France, France                NaN   

   host_response_rate  host_acceptance_rate host_is_superhost  \
0                 NaN                   NaN                 f   
1                 NaN                   NaN                 f   
2                 NaN                   NaN                 f   
3                 NaN                   NaN                 f   
4                 NaN                   NaN                 f   

   host_total_listings_count  ... minimum_nights maximum_nights  \
0                        1.0  ...              2           1125   
1                        1.0  ...              2           1125   
2                        1.0  ...              2           1125   
3                        1.0  ...              2           1125   
4                        1.0  ...              2           1125   

  review_scores_rating review_scores_accuracy review_scores_cleanliness  \
0                100.0                   10.0                      10.0   
1                100.0                   10.0                      10.0   
2                100.0                   10.0                      10.0   
3                100.0                   10.0                      10.0   
4                100.0                   10.0                      10.0   

   review_scores_checkin  review_scores_communication review_scores_location  \
0                   10.0                         10.0                   10.0   
1                   10.0                         10.0                   10.0   
2                   10.0                         10.0                   10.0   
3                   10.0                         10.0                   10.0   
4                   10.0                         10.0                   10.0   

  review_scores_value  instant_bookable  
0                10.0                 f  
1                10.0                 f  
2                10.0                 f  
3                10.0                 f  
4                10.0                 f  

[5 rows x 33 columns]

In [4]:
#view df tail

df.tail()

listing_id                                               name  \
279707    38338635  Appartement T2 neuf près du tram T3a Porte ...   
279708    38538692                          Cozy Studio in Montmartre   
279709    38683356            Nice and cosy mini-appartement in Paris   
279710    39659000  Charming apartment near Rue Saint Maur / Oberk...   
279711    40219504        Cosy apartment with view on Canal St Martin   

         host_id  host_since                 host_location host_response_time  \
279707  31161181  2015-04-13  Paris, Ile-de-France, France                NaN   
279708  10294858  2013-11-27  Paris, Ile-de-France, France                NaN   
279709   2238502  2012-04-27  Paris, Ile-de-France, France                NaN   
279710  38633695  2015-07-16  Paris, Ile-de-France, France                NaN   
279711   6955618  2013-06-17  Paris, Ile-de-France, France                NaN   

        host_response_rate  host_acceptance_rate host_is_superhost  \
279707                 NaN                   NaN                 f   
279708                 NaN                   NaN                 f   
279709                 NaN                   NaN                 f   
279710                 NaN                   NaN                 f   
279711                 NaN                   NaN                 f   

        host_total_listings_count  ... minimum_nights maximum_nights  \
279707                        1.0  ...              1              7   
279708                        1.0  ...              7             15   
279709                        1.0  ...              6             30   
279710                        1.0  ...              3             18   
279711                        1.0  ...              2              4   

       review_scores_rating review_scores_accuracy review_scores_cleanliness  \
279707                100.0                   10.0                      10.0   
279708                100.0                   10.0                      10.0   
279709                100.0                   10.0                      10.0   
279710                100.0                   10.0                      10.0   
279711                100.0                   10.0                      10.0   

        review_scores_checkin  review_scores_communication  \
279707                   10.0                         10.0   
279708                   10.0                         10.0   
279709                   10.0                         10.0   
279710                   10.0                         10.0   
279711                   10.0                         10.0   

       review_scores_location review_scores_value  instant_bookable  
279707                   10.0                10.0                 f  
279708                   10.0                10.0                 f  
279709                   10.0                10.0                 f  
279710                   10.0                10.0                 f  
279711                   10.0                10.0                 f  

[5 rows x 33 columns]

**The exploratory phase of the analysis consists in understanding your dataset in order to create a plan for cleaning, structuring, organizing and mapping data.**

In [5]:
#we observe there are 33 columns, but we can't see all of them with the head or tail function

df.columns

Index(['listing_id', 'name', 'host_id', 'host_since', 'host_location',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_total_listings_count',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'district', 'city', 'latitude', 'longitude', 'property_type',
       'room_type', 'accommodates', 'bedrooms', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'review_scores_rating',
       'review_scores_accuracy', 'review_scores_cleanliness',
       'review_scores_checkin', 'review_scores_communication',
       'review_scores_location', 'review_scores_value', 'instant_bookable'],
      dtype='object')

In [6]:
#let's read the listings data dictionary file to view the columns description
#we'll use the same method as above with pd read 

path2 =r'C:\\Users\Domnica Maria\Documents\1. PROJECTS\Maven-Vacation-Challenge\Airbnb Data\Listings_data_dictionary.csv'
df_dict = pd.read_csv(path2)

In [7]:
#view df dictionary 
#this will help us understand the fields content better in the exploratory phase and then later in the transformation phase

df_dict

Field  \
0                    listing_id   
1                          name   
2                       host_id   
3                    host_since   
4                 host_location   
5            host_response_time   
6            host_response_rate   
7          host_acceptance_rate   
8             host_is_superhost   
9     host_total_listings_count   
10         host_has_profile_pic   
11       host_identity_verified   
12                neighbourhood   
13                     district   
14                         city   
15                     latitude   
16                    longitude   
17                property_type   
18                    room_type   
19                 accommodates   
20                     bedrooms   
21                    amenities   
22                        price   
23               minimum_nights   
24               maximum_nights   
25         review_scores_rating   
26       review_scores_accuracy   
27    review_scores_cleanliness   
28        review_scores_checkin   
29  review_scores_communication   
30       review_scores_location   
31          review_scores_value   
32             instant_bookable   

                                          Description  
0                                          Listing ID  
1                                        Listing Name  
2                                             Host ID  
3                         Date the Host joined Airbnb  
4                    Location where the Host is based  
5      Estimate of how long the Host takes to respond  
6               Percentage of times the Host responds  
7   Percentage of times the Host accepts a booking...  
8   Binary field to determine if the Host is a Sup...  
9               Total listings the Host has in Airbnb  
10  Binary field to determine if the Host has a pr...  
11  Binary field to determine if the Host has a ve...  
12                     Neighborhood the Listing is in  
13                         District the Listing is in  
14                             City the Listing is in  
15                                 Listing's latitude  
16                                Listing's longitude  
17                   Type of property for the Listing  
18        Type of room type in Airbnb for the Listing  
19                     Guests the Listing accomodates  
20                            Bedrooms in the Listing  
21                     Amenities the Listing includes  
22         Listing price (in each country's currency)  
23                         Minimum nights per booking  
24                         Maximum nights per booking  
25              Listing's overall rating (out of 100)  
26  Listing's accuracy score based on what's promo...  
27            Listing's cleanliness score (out of 10)  
28    Listing's check-in experience score (out of 10)  
29  Listing's communication with the Host score (o...  
30  Listing's location score within the city (out ...  
31  Listing's value score relative to its price (o...  
32  Binary field to determine if the Listing can b...

In [8]:
#get more information by using the info function

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 279712 entries, 0 to 279711
Data columns (total 33 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   listing_id                   279712 non-null  int64  
 1   name                         279539 non-null  object 
 2   host_id                      279712 non-null  int64  
 3   host_since                   279547 non-null  object 
 4   host_location                278872 non-null  object 
 5   host_response_time           150930 non-null  object 
 6   host_response_rate           150930 non-null  float64
 7   host_acceptance_rate         166625 non-null  float64
 8   host_is_superhost            279547 non-null  object 
 9   host_total_listings_count    279547 non-null  float64
 10  host_has_profile_pic         279547 non-null  object 
 11  host_identity_verified       279547 non-null  object 
 12  neighbourhood                279712 non-null  object 
 13 

We know we have 279712 records in this dataset and 33 columns. 

**Statistics summary**

In [9]:
#get a statistical summary of each column, such as count, column mean value, column standard deviation, etc. 
#We use the describe function

df.describe()

listing_id       host_id  host_response_rate  host_acceptance_rate  \
count  2.797120e+05  2.797120e+05       150930.000000         166625.000000   
mean   2.638196e+07  1.081658e+08            0.865939              0.827168   
std    1.442576e+07  1.108570e+08            0.283744              0.289202   
min    2.577000e+03  1.822000e+03            0.000000              0.000000   
25%    1.384462e+07  1.720656e+07            0.900000              0.780000   
50%    2.767098e+07  5.826911e+07            1.000000              0.980000   
75%    3.978485e+07  1.832853e+08            1.000000              1.000000   
max    4.834353e+07  3.901874e+08            1.000000              1.000000   

       host_total_listings_count       latitude      longitude   accommodates  \
count              279547.000000  279712.000000  279712.000000  279712.000000   
mean                   24.581612      18.761862      12.595075       3.288736   
std                   284.041143      32.560343      73.081309       2.133379   
min                     0.000000     -34.264400     -99.339630       0.000000   
25%                     1.000000     -22.964390     -43.198040       2.000000   
50%                     1.000000      40.710785       2.382780       2.000000   
75%                     4.000000      41.908610      28.986730       4.000000   
max                  7235.000000      48.904910     151.339810      16.000000   

            bedrooms          price  minimum_nights  maximum_nights  \
count  250277.000000  279712.000000   279712.000000    2.797120e+05   
mean        1.515509     608.792737        8.050967    2.755860e+04   
std         1.153080    3441.826611       31.518946    7.282875e+06   
min         1.000000       0.000000        1.000000    1.000000e+00   
25%         1.000000      75.000000        1.000000    4.500000e+01   
50%         1.000000     150.000000        2.000000    1.125000e+03   
75%         2.000000     474.000000        5.000000    1.125000e+03   
max        50.000000  625216.000000     9999.000000    2.147484e+09   

       review_scores_rating  review_scores_accuracy  \
count         188307.000000           187999.000000   
mean              93.405195                9.565476   
std               10.070437                0.990878   
min               20.000000                2.000000   
25%               91.000000                9.000000   
50%               96.000000               10.000000   
75%              100.000000               10.000000   
max              100.000000               10.000000   

       review_scores_cleanliness  review_scores_checkin  \
count              188047.000000          187941.000000   
mean                    9.312869               9.701534   
std                     1.146072               0.867434   
min                     2.000000               2.000000   
25%                     9.000000              10.000000   
50%                    10.000000              10.000000   
75%                    10.000000              10.000000   
max                    10.000000              10.000000   

       review_scores_communication  review_scores_location  \
count                188025.000000           187937.000000   
mean                      9.698593                9.633994   
std                       0.886884                0.833234   
min                       2.000000                2.000000   
25%                      10.000000                9.000000   
50%                      10.000000               10.000000   
75%                      10.000000               10.000000   
max                      10.000000               10.000000   

       review_scores_value  
count        187927.000000  
mean              9.335364  
std               1.042625  
min               2.000000  
25%               9.000000  
50%              10.000000  
75%              10.000000  
max              10.000000

In [10]:
#we observe the various statistics summary, excluding NaN (Not a Number) values
#however, these are shown only for numeric-type columns such as int and float
#use the following function to include all columns 

df.describe(include="all")

listing_id                 name       host_id  host_since  \
count   2.797120e+05               279539  2.797120e+05      279547   
unique           NaN               265861           NaN        4240   
top              NaN  home away from home           NaN  2019-09-02   
freq             NaN                   98           NaN         710   
mean    2.638196e+07                  NaN  1.081658e+08         NaN   
std     1.442576e+07                  NaN  1.108570e+08         NaN   
min     2.577000e+03                  NaN  1.822000e+03         NaN   
25%     1.384462e+07                  NaN  1.720656e+07         NaN   
50%     2.767098e+07                  NaN  5.826911e+07         NaN   
75%     3.978485e+07                  NaN  1.832853e+08         NaN   
max     4.834353e+07                  NaN  3.901874e+08         NaN   

                       host_location host_response_time  host_response_rate  \
count                         278872             150930       150930.000000   
unique                          7159                  4                 NaN   
top     Paris, Ile-de-France, France     within an hour                 NaN   
freq                           47794              83464                 NaN   
mean                             NaN                NaN            0.865939   
std                              NaN                NaN            0.283744   
min                              NaN                NaN            0.000000   
25%                              NaN                NaN            0.900000   
50%                              NaN                NaN            1.000000   
75%                              NaN                NaN            1.000000   
max                              NaN                NaN            1.000000   

        host_acceptance_rate host_is_superhost  host_total_listings_count  \
count          166625.000000            279547              279547.000000   
unique                   NaN                 2                        NaN   
top                      NaN                 f                        NaN   
freq                     NaN            229294                        NaN   
mean                0.827168               NaN                  24.581612   
std                 0.289202               NaN                 284.041143   
min                 0.000000               NaN                   0.000000   
25%                 0.780000               NaN                   1.000000   
50%                 0.980000               NaN                   1.000000   
75%                 1.000000               NaN                   4.000000   
max                 1.000000               NaN                7235.000000   

        ... minimum_nights maximum_nights review_scores_rating  \
count   ...  279712.000000   2.797120e+05        188307.000000   
unique  ...            NaN            NaN                  NaN   
top     ...            NaN            NaN                  NaN   
freq    ...            NaN            NaN                  NaN   
mean    ...       8.050967   2.755860e+04            93.405195   
std     ...      31.518946   7.282875e+06            10.070437   
min     ...       1.000000   1.000000e+00            20.000000   
25%     ...       1.000000   4.500000e+01            91.000000   
50%     ...       2.000000   1.125000e+03            96.000000   
75%     ...       5.000000   1.125000e+03           100.000000   
max     ...    9999.000000   2.147484e+09           100.000000   

       review_scores_accuracy review_scores_cleanliness  \
count           187999.000000             188047.000000   
unique                    NaN                       NaN   
top                       NaN                       NaN   
freq                      NaN                       NaN   
mean                 9.565476                  9.312869   
std                  0.990878                  1.146072   
min                  2.000000                  2.000000   
25%               

In [11]:
#statics summary for object-typed attributes includes:
#how many unique values, which is the top value and the frequency of top value in the object-typed columns
#some values show as "NaN" and this is because those numbers are not available regarding a particular column type 

**Observations after statistics summary:**

* a lot of hosts joined in September 2019
* the most common host_location is Paris 
* hosts respond quickly, the majority of them within an hour, resulting in an overall good response, this being highlighted in the response rate as well
* host_response_rate shows that even if the min is 0, just 25% of records are under 0.9, which is a really good rate, it shows responsability and seriousness from hosts
* host_acceptance_rate has an overall good rate, showing that only 25% of hosts are more selective (rate 0.78) and the rest accept approximately all (rate 0.98, respectively 1)
* most of the hosts are not superhosts 
* 50% of hosts have only 1 listing, max value 7235 which we need to investigate (potential outlier)
* 75% of listings can accomodate up to 4 guests and the max number of guests possible to accomodate is 16 in some cases
* 75% of accomodations offer up to 2 bedrooms and the max value is 50 which we need to investigate to check if outlier or not
* price recorded in some cases is 0 as observed with the min, which would need to be removed as our goal is to analyze the price influencing factors and if the listing doesn't have a price, these records are redundant 
* max value for minimum_nights is 9999 which looks like an outlier, we'll investigate closer in the next steps
* min review score rating is 20 with but the majority of reviews are above 90, which shows good feedback about listings overall
* similar results observed in all review related columns, strengthening the facts 

**Evaluating for missing data**

Next step is to evaluate the missing values. We use the one of the 2 methods to identify these:

* .isnull()

* .notnull()

The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [12]:
missing_data = df.isnull()
missing_data

listing_id   name  host_id  host_since  host_location  \
0            False  False    False       False          False   
1            False  False    False       False          False   
2            False  False    False       False          False   
3            False  False    False       False          False   
4            False  False    False       False          False   
...            ...    ...      ...         ...            ...   
279707       False  False    False       False          False   
279708       False  False    False       False          False   
279709       False  False    False       False          False   
279710       False  False    False       False          False   
279711       False  False    False       False          False   

        host_response_time  host_response_rate  host_acceptance_rate  \
0                     True                True                  True   
1                     True                True                  True   
2                     True                True                  True   
3                     True                True                  True   
4                     True                True                  True   
...                    ...                 ...                   ...   
279707                True                True                  True   
279708                True                True                  True   
279709                True                True                  True   
279710                True                True                  True   
279711                True                True                  True   

        host_is_superhost  host_total_listings_count  ...  minimum_nights  \
0                   False                      False  ...           False   
1                   False                      False  ...           False   
2                   False                      False  ...           False   
3                   False                      False  ...           False   
4                   False                      False  ...           False   
...                   ...                        ...  ...             ...   
279707              False                      False  ...           False   
279708              False                      False  ...           False   
279709              False                      False  ...           False   
279710              False                      False  ...           False   
279711              False                      False  ...           False   

        maximum_nights  review_scores_rating  review_scores_accuracy  \
0                False                 False                   False   
1                False                 False                   False   
2                False                 False                   False   
3                False                 False                   False   
4                False                 False                   False   
...                ...                   ...                     ...   
279707           False                 False                   False   
279708           False                 False                   False   
279709           False                 False                   False   
279710           False                 False                   False   
279711           False                 False                   False   

        review_scores_cleanliness  review_scores_checkin  \
0                           False                  False   
1                           False                  False   
2                           False                  False   
3                           False                  False   
4                           False                  False   
...                           ...                    ...   
279707                      False                  False   
279708                      False                  False   
279709                      Fals

**Count missing values in each column**

Using a for loop in Python, we can quickly figure out the number of missing values in each column. 

As mentioned above, "True" represents a missing value, "False" means the value is present in the dataset. 

In the body of the for loop the method ".value_counts()" counts the number of "True" values.

In [13]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

listing_id
False    279712
Name: listing_id, dtype: int64

name
False    279539
True        173
Name: name, dtype: int64

host_id
False    279712
Name: host_id, dtype: int64

host_since
False    279547
True        165
Name: host_since, dtype: int64

host_location
False    278872
True        840
Name: host_location, dtype: int64

host_response_time
False    150930
True     128782
Name: host_response_time, dtype: int64

host_response_rate
False    150930
True     128782
Name: host_response_rate, dtype: int64

host_acceptance_rate
False    166625
True     113087
Name: host_acceptance_rate, dtype: int64

host_is_superhost
False    279547
True        165
Name: host_is_superhost, dtype: int64

host_total_listings_count
False    279547
True        165
Name: host_total_listings_count, dtype: int64

host_has_profile_pic
False    279547
True        165
Name: host_has_profile_pic, dtype: int64

host_identity_verified
False    279547
True        165
Name: host_identity_verified, dtype: int64

neig

Based on the above output, the following include null values: 

* columns with categorical values:
  * name 
  * host_since
  * host_location
  * host_response_time
  * host_is_superhost
  * host_has_profile_pic
  * host_identity_verified
  * district
  
  
* columns with numerical values:
  * host_response_rate
  * host_acceptance_rate
  * host_total_listings_count
  * bedrooms
  * review_scores_rating
  * review_scores_accuracy
  * review_scores_cleanliness
  * review_scores_checkin
  * review_scores_communication
  * review_scores_location
  * review_scores_value

18 columns out of 33 contain null values

Keep in mind that based on the statistics summary, we know that the following columns have 0 values, not counted as missing values, but in context represent null values: 

* accomodates
* price 

**Deal with missing values**

* drop data
  * drop the whole row
  * drop the whole column
* replace data (depending on data type)
  * replace it by mean if numerical value
  * replace it by frequency if categorical value
  * replace it based on other functions

We'll address first the **categorical values** missing.

In [14]:
#first on the list is the name columns
#we know already from the statistics summary that the most frequent value in this column is "home away from home"
#by using value_counts() I'm checking what other values are most common 

df['name'].value_counts()

home away from home                                   98
Home sweet home                                       80
Private Room                                          59
Perfect Loft great location                           55
New!  La Chada Night Market studio 2PPL near MRT      48
                                                      ..
City Apartment. Close to china town + FREE PARKING     1
The Square 101 Superior studio 305 (Small Balcony)     1
ROOM WITH BREAKFAST                                    1
Historic Brownstone in Harlem                          1
condo cani 701                                         1
Name: name, Length: 265861, dtype: int64

In [15]:
#replace missing values with "home away from home"

df['name'] = df['name'].replace(np.nan, 'home away from home')

In [16]:
#since we had 173 missing values in name column, after the change we should have 271 records for "home away from home" 
#verifying with value_counts()

df['name'].value_counts()

home away from home                                   271
Home sweet home                                        80
Private Room                                           59
Perfect Loft great location                            55
New!  La Chada Night Market studio 2PPL near MRT       48
                                                     ... 
City Apartment. Close to china town + FREE PARKING      1
The Square 101 Superior studio 305 (Small Balcony)      1
ROOM WITH BREAKFAST                                     1
Historic Brownstone in Harlem                           1
condo cani 701                                          1
Name: name, Length: 265861, dtype: int64

In [17]:
#we move onto the next column on the list host_since with 165 missing values

df['host_since'].value_counts()

2019-09-02    710
2018-02-24    463
2013-07-16    463
2016-09-29    415
2015-12-16    368
             ... 
2009-05-21      1
2009-07-28      1
2010-01-26      1
2009-12-14      1
2009-09-09      1
Name: host_since, Length: 4240, dtype: int64

What I'm thinking is that I don't want to lose the records without date, but I don't want to replace the NaN with the most frequent value, because it will affect the analysis later. I think the best option at this point is replacing it with today's date. 

After analyzing the main influencing factors of price, I might consider dropping these records with NaN/today's date in case I use host_since for this purpose. We'll review this later based on analysis. 

In [18]:
#first import date library 

from datetime import date

In [19]:
#get today's date 2021-04-21

today = str(date.today())
print(today)

2021-04-25


In [20]:
#replace NaN with today's date

df['host_since'] = df['host_since'].replace(np.nan,today)

In [21]:
#next we look at host_location with 840 missing values

df['host_location'].value_counts()

Paris, Ile-de-France, France                         47794
New York, New York, United States                    24040
Rome, Lazio, Italy                                   20138
Cape Town, Western Cape, South Africa                13602
Rio de Janeiro, State of Rio de Janeiro, Brazil      13211
                                                     ...  
Puiseaux, Centre, France                                 1
Chabournay, Poitou-Charentes, France                     1
BTS Wong Wien Yai                                        1
Clairefontaine-en-Yvelines, Ile-de-France, France        1
València, Comunidad Valenciana, Spain                 1
Name: host_location, Length: 7159, dtype: int64

Same as host_since, I don't want to drop the records and lose potential good data. I will not replace with the most frequent value, rather with 'unknown'. 

We'll reconsider this step if we extend the potential price influencing factors to host characteristics, same as host_since.

In [22]:
#replace NaN with 'unknown'

df['host_location'] = df['host_location'].replace(np.nan,'unknown')

In [23]:
#next we look at host_response_time, 128782 missing values

df['host_response_time'].value_counts()

within an hour        83464
within a few hours    28891
within a day          23425
a few days or more    15150
Name: host_response_time, dtype: int64

This step is important because almost half of the records are missing values, more specifically 128782. 

We know from statistics summary that only 25% of host_response_rate is under 0.9, which gives me confidence in the decision. We will replace the NaN with the most frequent value 'within an hour'.

In [24]:
#replace NaN with 'within an hour'

df['host_response_time'] = df['host_response_time'].replace(np.nan,'within an hour')

In [25]:
#since we had 128782 missing values, now we should have 212246 rows for 'within an hour'

df['host_response_time'].value_counts()

within an hour        212246
within a few hours     28891
within a day           23425
a few days or more     15150
Name: host_response_time, dtype: int64

In [26]:
#next host_is_superhost, missing 165 values 

df['host_is_superhost'].value_counts()

f    229294
t     50253
Name: host_is_superhost, dtype: int64

In [27]:
#we observe that a vast majority is not a superhost, therefore replacing NaN with the most frequent value 

df['host_is_superhost'] = df['host_is_superhost'].replace(np.nan,'f')

In [28]:
#after change, we should have 229459 'f' values 

df['host_is_superhost'].value_counts()

f    229459
t     50253
Name: host_is_superhost, dtype: int64

In [29]:
#next host_has_profile_pic, 165 missing values

df['host_has_profile_pic'].value_counts()

t    278631
f       916
Name: host_has_profile_pic, dtype: int64

In [30]:
#we'll replace NaN with the most frequent value 't'

df['host_has_profile_pic'] = df['host_has_profile_pic'].replace(np.nan,'t')

In [31]:
#we should have now a total of 278796 't' values

df['host_has_profile_pic'].value_counts()

t    278796
f       916
Name: host_has_profile_pic, dtype: int64

In [32]:
#next host_identity_verified, 165 missing values

df['host_identity_verified'].value_counts()

t    201191
f     78356
Name: host_identity_verified, dtype: int64

In [33]:
#we'll replace NaN with most frequent value 't'

df['host_identity_verified'] = df['host_identity_verified'].replace(np.nan,'t')

In [34]:
#we should have now 201356 't' values


df['host_identity_verified'].value_counts()

t    201356
f     78356
Name: host_identity_verified, dtype: int64

In [35]:
#last categorical column is district with 242700 missing values

df['district'].value_counts()

Manhattan        16553
Brooklyn         14474
Queens            4704
Bronx              992
Staten Island      289
Name: district, dtype: int64

In [36]:
#we observe that the district is specified only for US based listings
#we'll replace NaN with 'NA' for not applicable 

df['district'] = df['district'].replace(np.nan,'NA')

In [37]:
#we should see now NA in value_counts() output 

df['district'].value_counts()

NA               242700
Manhattan         16553
Brooklyn          14474
Queens             4704
Bronx               992
Staten Island       289
Name: district, dtype: int64

We'll address now the **numerical values** missing.

In [38]:
#first on the list is host_response_rate 
#we'll replace with the 128782 missing values with the mean value as this coloumn contains numerical values 

df['host_response_rate'].value_counts()

1.00    99334
0.00    10640
0.90     4982
0.50     3639
0.80     3085
        ...  
0.45        4
0.15        3
0.21        3
0.24        1
0.01        1
Name: host_response_rate, Length: 91, dtype: int64

In [39]:
#get mean value

mean_hrr = df['host_response_rate'].mean()

#format mean_hrr

mean_hrr = float("{:.2f}".format(mean_hrr))

print("Mean value for host_response_rate is", mean_hrr)

Mean value for host_response_rate is 0.87


In [40]:
#replace NaN with mean_hrr 0.87

df['host_response_rate'] = df['host_response_rate'].replace(np.nan,mean_hrr)

In [41]:
#considering we had 128782 missing values, which we should see now with 0.87

df['host_response_rate'].value_counts()

0.87    129218
1.00     99334
0.00     10640
0.90      4982
0.50      3639
         ...  
0.45         4
0.15         3
0.21         3
0.24         1
0.01         1
Name: host_response_rate, Length: 91, dtype: int64

In [42]:
#moving next to host_acceptance_rate, 113087 missing values

df['host_acceptance_rate'].value_counts()

1.00    75555
0.00    11431
0.99     7227
0.50     5445
0.98     5085
        ...  
0.05       28
0.01       25
0.21       23
0.02       14
0.16        4
Name: host_acceptance_rate, Length: 101, dtype: int64

In [43]:
#we'll action it similiar as above by replacing NaN with mean value
#get mean value first 

mean_har = df['host_acceptance_rate'].mean()

#format mean_har

mean_har = float("{:.2f}".format(mean_har))

print("Mean value for host_acceptance_rate is", mean_har)

Mean value for host_acceptance_rate is 0.83


In [44]:
#replace NaN with mean_har 0.83

df['host_acceptance_rate'] = df['host_acceptance_rate'].replace(np.nan,mean_har)

In [45]:
#check the transformation

df['host_acceptance_rate'].value_counts()

0.83    115106
1.00     75555
0.00     11431
0.99      7227
0.50      5445
         ...  
0.05        28
0.01        25
0.21        23
0.02        14
0.16         4
Name: host_acceptance_rate, Length: 101, dtype: int64

In [46]:
#next is host_total_listings_count with 165 missing values

df['host_total_listings_count'].value_counts()

1.0       117559
2.0        35778
0.0        33265
3.0        17230
4.0        11216
           ...  
172.0          1
1360.0         1
276.0          1
852.0          1
653.0          1
Name: host_total_listings_count, Length: 206, dtype: int64

In [47]:
#check mean value 
df['host_total_listings_count'].mean()

24.581612394337984

In [48]:
#we know from statistics summary that %50 of values represent 1 listing per host, making this a potentail good solution
#mean value is not a good choice to pursue due to high value generated by potential outliers in dataset
#we remember the max value is 7235 

df['host_total_listings_count'] = df['host_total_listings_count'].replace(np.nan,1.0)

In [49]:
#we should see now 117724 records for 1 listing

df['host_total_listings_count'].value_counts()

1.0       117724
2.0        35778
0.0        33265
3.0        17230
4.0        11216
           ...  
172.0          1
1360.0         1
276.0          1
852.0          1
653.0          1
Name: host_total_listings_count, Length: 206, dtype: int64

In [50]:
#next is bedrooms with 29435 missing values

df['bedrooms'].value_counts()

1.0     170163
2.0      51382
3.0      18525
4.0       6579
5.0       2106
6.0        701
7.0        246
10.0       150
8.0        124
9.0         79
12.0        38
11.0        32
50.0        22
20.0        22
16.0        16
14.0        15
15.0        12
40.0        10
13.0         7
22.0         5
23.0         4
24.0         4
30.0         4
18.0         4
19.0         3
33.0         3
17.0         3
38.0         2
39.0         2
46.0         2
35.0         2
31.0         2
21.0         2
41.0         1
42.0         1
26.0         1
48.0         1
34.0         1
25.0         1
Name: bedrooms, dtype: int64

In [51]:
#get mean 

mean_b = df['bedrooms'].mean()

#format mean_b

mean_b = float("{:.2f}".format(mean_b))

print("Mean value for bedrooms is", mean_b)

Mean value for bedrooms is 1.52


In [52]:
#replace NaN with mean_b 1.52

df['bedrooms'] = df['bedrooms'].replace(np.nan,mean_b)

In [53]:
#we should see now 29435 values for 1.52

df['bedrooms'].value_counts()

1.00     170163
2.00      51382
1.52      29435
3.00      18525
4.00       6579
5.00       2106
6.00        701
7.00        246
10.00       150
8.00        124
9.00         79
12.00        38
11.00        32
20.00        22
50.00        22
16.00        16
14.00        15
15.00        12
40.00        10
13.00         7
22.00         5
30.00         4
24.00         4
23.00         4
18.00         4
17.00         3
19.00         3
33.00         3
21.00         2
39.00         2
38.00         2
46.00         2
31.00         2
35.00         2
34.00         1
41.00         1
48.00         1
26.00         1
42.00         1
25.00         1
Name: bedrooms, dtype: int64

In [54]:
#next is review_scores_rating with 91405 missing values 

df['review_scores_rating'].value_counts()

100.0    57458
98.0     13616
97.0     12425
96.0     12261
93.0     10995
         ...  
36.0         1
61.0         1
31.0         1
44.0         1
49.0         1
Name: review_scores_rating, Length: 63, dtype: int64

In [55]:
#get mean

mean_rsr = df['review_scores_rating'].mean()

#format mean_rsr

mean_rsr = float("{:.2f}".format(mean_rsr))

print("Mean value for review_scores_rating is", mean_rsr)

Mean value for review_scores_rating is 93.41


In [56]:
#replace NaN with mean_rsr 93.41

df['review_scores_rating'] = df['review_scores_rating'].replace(np.nan,mean_rsr)

In [57]:
#we should see now 91405 values for 93.41

df['review_scores_rating'].value_counts()

93.41     91405
100.00    57458
98.00     13616
97.00     12425
96.00     12261
          ...  
36.00         1
31.00         1
61.00         1
44.00         1
49.00         1
Name: review_scores_rating, Length: 64, dtype: int64

In [58]:
#next we look at review_scores_accuracy with 91713 missing values 

df['review_scores_accuracy'].value_counts()

10.0    136390
9.0      37176
8.0       8712
6.0       1974
7.0       1795
2.0       1152
4.0        527
5.0        240
3.0         33
Name: review_scores_accuracy, dtype: int64

In [59]:
#get mean 

mean_rsa = df['review_scores_accuracy'].mean()

#format mean_rsa

mean_rsa = float("{:.2f}".format(mean_rsa))

print("Mean value for review_scores_accuracy is", mean_rsa)

Mean value for review_scores_accuracy is 9.57


In [60]:
#replace NaN with mean_rsa 9.57

df['review_scores_accuracy'] = df['review_scores_accuracy'].replace(np.nan,mean_rsa)

In [61]:
#we should see now 91713 values of 9.57

df['review_scores_accuracy'].value_counts()

10.00    136390
9.57      91713
9.00      37176
8.00       8712
6.00       1974
7.00       1795
2.00       1152
4.00        527
5.00        240
3.00         33
Name: review_scores_accuracy, dtype: int64

In [62]:
#next is review_scores_cleanliness with 91655 missing values 

df['review_scores_cleanliness'].value_counts()

10.0    109029
9.0      52107
8.0      17148
7.0       3810
6.0       3209
2.0       1271
4.0        853
5.0        541
3.0         79
Name: review_scores_cleanliness, dtype: int64

In [63]:
#get mean 

mean_rscl = df['review_scores_cleanliness'].mean()

#format mean_rscl

mean_rscl = float("{:.2f}".format(mean_rscl))

print("Mean value for review_scores_cleanliness is", mean_rscl)

Mean value for review_scores_cleanliness is 9.31


In [64]:
#replace NaN with mean_rscl 9.31

df['review_scores_cleanliness'] = df['review_scores_cleanliness'].replace(np.nan,mean_rscl)

In [65]:
#we should see now 91655 value of 9.31

df['review_scores_cleanliness'].value_counts()

10.00    109029
9.31      91665
9.00      52107
8.00      17148
7.00       3810
6.00       3209
2.00       1271
4.00        853
5.00        541
3.00         79
Name: review_scores_cleanliness, dtype: int64

In [66]:
#next is review_scores_checkin with 91771 values 

df['review_scores_checkin'].value_counts()

10.0    152997
9.0      25292
8.0       5685
6.0       1381
7.0       1135
2.0        963
4.0        327
5.0        145
3.0         16
Name: review_scores_checkin, dtype: int64

In [67]:
#get mean 

mean_rsch = df['review_scores_checkin'].mean()

#format mean_rsch

mean_rsch = float("{:.2f}".format(mean_rsch))

print("Mean value for review_scores_checkin is", mean_rsch)

Mean value for review_scores_checkin is 9.7


In [68]:
#replace NaN with mean_rsch 9.7

df['review_scores_checkin'] = df['review_scores_checkin'].replace(np.nan,mean_rsch)

In [69]:
#we should see now 91771 values for 9.7

df['review_scores_checkin'].value_counts()

10.0    152997
9.7      91771
9.0      25292
8.0       5685
6.0       1381
7.0       1135
2.0        963
4.0        327
5.0        145
3.0         16
Name: review_scores_checkin, dtype: int64

In [70]:
#next is review_scores_communication with 91687 missing values

df['review_scores_communication'].value_counts()

10.0    153549
9.0      24572
8.0       5668
6.0       1434
7.0       1229
2.0       1020
4.0        372
5.0        159
3.0         22
Name: review_scores_communication, dtype: int64

In [71]:
#get mean 

mean_rsco = df['review_scores_communication'].mean()

#format mean_rsco

mean_rsco = float("{:.2f}".format(mean_rsco))

print("Mean value for review_scores_communication is", mean_rsco)

Mean value for review_scores_communication is 9.7


In [72]:
#replace NaN with 9.70

df['review_scores_communication'] = df['review_scores_communication'].replace(np.nan,9.70)

In [73]:
#we should see now 91687 values for 9.70

df['review_scores_communication'].value_counts()

10.0    153549
9.7      91687
9.0      24572
8.0       5668
6.0       1434
7.0       1229
2.0       1020
4.0        372
5.0        159
3.0         22
Name: review_scores_communication, dtype: int64

In [74]:
#next is review_scores_location with 91775 missing values

df['review_scores_location'].value_counts()

10.0    140495
9.0      35350
8.0       8465
6.0       1480
7.0       1151
2.0        635
4.0        234
5.0        120
3.0          7
Name: review_scores_location, dtype: int64

In [75]:
#get mean 

mean_rsl = df['review_scores_location'].mean()

#format mean_rsl

mean_rsl = float("{:.2f}".format(mean_rsl)) 

print("Mean value for review_scores_location is", mean_rsl)

Mean value for review_scores_location is 9.63


In [76]:
#replace NaN with mean_rsl 9.63

df['review_scores_location'] = df['review_scores_location'].replace(np.nan,mean_rsl)

In [77]:
#we should see now 91775 values for 9.63

df['review_scores_location'].value_counts()

10.00    140495
9.63      91775
9.00      35350
8.00       8465
6.00       1480
7.00       1151
2.00        635
4.00        234
5.00        120
3.00          7
Name: review_scores_location, dtype: int64

In [78]:
#last column on the list is review_scores_value with 91785 missing values 

df['review_scores_value'].value_counts()

10.0    102530
9.0      63670
8.0      14588
6.0       2602
7.0       2497
2.0       1095
4.0        605
5.0        306
3.0         34
Name: review_scores_value, dtype: int64

In [79]:
#get mean 

mean_rsv = df['review_scores_value'].mean()

mean_rsv = float("{:.2f}".format(mean_rsv))

print("Mean value for review_scores_value is", mean_rsv)

Mean value for review_scores_value is 9.34


In [80]:
#replace NaN with mean_rsv 9.34

df['review_scores_value'] = df['review_scores_value'].replace(np.nan,mean_rsv)

In [81]:
#we should see now 91785 values for 9.34

df['review_scores_value'].value_counts()

10.00    102530
9.34      91785
9.00      63670
8.00      14588
6.00       2602
7.00       2497
2.00       1095
4.00        605
5.00        306
3.00         34
Name: review_scores_value, dtype: int64

All missing values have been replaced accordingly. Let's run the loop again for a final check. 

In [82]:
missing_data = df.isnull()
missing_data

listing_id   name  host_id  host_since  host_location  \
0            False  False    False       False          False   
1            False  False    False       False          False   
2            False  False    False       False          False   
3            False  False    False       False          False   
4            False  False    False       False          False   
...            ...    ...      ...         ...            ...   
279707       False  False    False       False          False   
279708       False  False    False       False          False   
279709       False  False    False       False          False   
279710       False  False    False       False          False   
279711       False  False    False       False          False   

        host_response_time  host_response_rate  host_acceptance_rate  \
0                    False               False                 False   
1                    False               False                 False   
2                    False               False                 False   
3                    False               False                 False   
4                    False               False                 False   
...                    ...                 ...                   ...   
279707               False               False                 False   
279708               False               False                 False   
279709               False               False                 False   
279710               False               False                 False   
279711               False               False                 False   

        host_is_superhost  host_total_listings_count  ...  minimum_nights  \
0                   False                      False  ...           False   
1                   False                      False  ...           False   
2                   False                      False  ...           False   
3                   False                      False  ...           False   
4                   False                      False  ...           False   
...                   ...                        ...  ...             ...   
279707              False                      False  ...           False   
279708              False                      False  ...           False   
279709              False                      False  ...           False   
279710              False                      False  ...           False   
279711              False                      False  ...           False   

        maximum_nights  review_scores_rating  review_scores_accuracy  \
0                False                 False                   False   
1                False                 False                   False   
2                False                 False                   False   
3                False                 False                   False   
4                False                 False                   False   
...                ...                   ...                     ...   
279707           False                 False                   False   
279708           False                 False                   False   
279709           False                 False                   False   
279710           False                 False                   False   
279711           False                 False                   False   

        review_scores_cleanliness  review_scores_checkin  \
0                           False                  False   
1                           False                  False   
2                           False                  False   
3                           False                  False   
4                           False                  False   
...                           ...                    ...   
279707                      False                  False   
279708                      False                  False   
279709                      Fals

In [83]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

listing_id
False    279712
Name: listing_id, dtype: int64

name
False    279712
Name: name, dtype: int64

host_id
False    279712
Name: host_id, dtype: int64

host_since
False    279712
Name: host_since, dtype: int64

host_location
False    279712
Name: host_location, dtype: int64

host_response_time
False    279712
Name: host_response_time, dtype: int64

host_response_rate
False    279712
Name: host_response_rate, dtype: int64

host_acceptance_rate
False    279712
Name: host_acceptance_rate, dtype: int64

host_is_superhost
False    279712
Name: host_is_superhost, dtype: int64

host_total_listings_count
False    279712
Name: host_total_listings_count, dtype: int64

host_has_profile_pic
False    279712
Name: host_has_profile_pic, dtype: int64

host_identity_verified
False    279712
Name: host_identity_verified, dtype: int64

neighbourhood
False    279712
Name: neighbourhood, dtype: int64

district
False    279712
Name: district, dtype: int64

city
False    279712
Name: city, dtype: int6

Great! Exactly what I was expecting to see, no more missing values found. 

There's one more step to do in this phase and that is dropping the rows with price 0, because this is our target variable and it can't be 0 for analysis. 

In [84]:
#let's view first how many records we'll be losing during this step
#create a df to view the boolean values of the next statement 

df_price_zero = df['price'] == 0

#check how many rows have value 0 for price 

df_price_zero.value_counts()

False    279599
True        113
Name: price, dtype: int64

In [85]:
#drop rows with value 0 for price

df.drop(df.index[df['price'] == 0], inplace = True)

In [86]:
#check if any rows left with 0 by executing the above statement again 

df_price_zero = df['price'] == 0 

df_price_zero.value_counts()

False    279599
Name: price, dtype: int64

No more rows with price 0. Let's reset the df index and see how many rows we have now in total left.

In [87]:
#reset df index

df = df.reset_index(drop=True)

In [88]:
#check how many rows left

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 279599 entries, 0 to 279598
Data columns (total 33 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   listing_id                   279599 non-null  int64  
 1   name                         279599 non-null  object 
 2   host_id                      279599 non-null  int64  
 3   host_since                   279599 non-null  object 
 4   host_location                279599 non-null  object 
 5   host_response_time           279599 non-null  object 
 6   host_response_rate           279599 non-null  float64
 7   host_acceptance_rate         279599 non-null  float64
 8   host_is_superhost            279599 non-null  object 
 9   host_total_listings_count    279599 non-null  float64
 10  host_has_profile_pic         279599 non-null  object 
 11  host_identity_verified       279599 non-null  object 
 12  neighbourhood                279599 non-null  object 
 13 

We are working now with 279598 records. 

Next step in our transformation phase is converting the data types to the proper format.

In [89]:
#view df data types

df.dtypes

listing_id                       int64
name                            object
host_id                          int64
host_since                      object
host_location                   object
host_response_time              object
host_response_rate             float64
host_acceptance_rate           float64
host_is_superhost               object
host_total_listings_count      float64
host_has_profile_pic            object
host_identity_verified          object
neighbourhood                   object
district                        object
city                            object
latitude                       float64
longitude                      float64
property_type                   object
room_type                       object
accommodates                     int64
bedrooms                       float64
amenities                       object
price                            int64
minimum_nights                   int64
maximum_nights                   int64
review_scores_rating     

Let's plan our next steps in the process of data type changes. This way we don't miss anything. 

Some observations are for the following step which is to normalize data, more specifically to standardize data.

In [90]:
# listing_id                       int64  - ok, no change needed 
# name                            object  - ok 
# host_id                          int64  - ok
# host_since                      object  - change to datatime 
# host_location                   object  - ok 
# host_response_time              object  - ok
# host_response_rate             float64  - ok
# host_acceptance_rate           float64  - ok
# host_is_superhost               object  - change to int, transform t=1 and f=0
# host_total_listings_count      float64  - change to int, no floating values 
# host_has_profile_pic            object  - change to int, transform t=1 and f=0
# host_identity_verified          object  - change to int, transform t=1 and f=0
# neighbourhood                   object  - ok
# district                        object  - ok
# city                            object  - ok
# latitude                       float64  - ok
# longitude                      float64  - ok
# property_type                   object  - ok
# room_type                       object  - ok
# accommodates                     int64  - ok
# bedrooms                       float64  - ok
# amenities                       object  - (see below) 
# price                            int64  - (see below)
# minimum_nights                   int64  - ok
# maximum_nights                   int64  - ok
# review_scores_rating           float64  - ok
# review_scores_accuracy         float64  - ok
# review_scores_cleanliness      float64  - ok
# review_scores_checkin          float64  - ok
# review_scores_communication    float64  - ok
# review_scores_location         float64  - ok
# review_scores_value            float64  - ok
# instant_bookable                object  - change to int, transform t=1 and f=0

In [91]:
#first dtype change for column host_since from object to datetime 

df['host_since'] = pd.to_datetime(df['host_since'])

In [92]:
#check dtype after change 

df['host_since'].dtypes

dtype('<M8[ns]')

In [93]:
#verification step

np.dtype('datetime64[ns]') == np.dtype('<M8[ns]') 

True

In [94]:
#we'll use replace method instead of get_dummies to 

df['host_is_superhost'] = df['host_is_superhost'].replace('f',0)
df['host_is_superhost'] = df['host_is_superhost'].replace('t',1)

In [95]:
#verification step

df['host_is_superhost'].value_counts()

0    229350
1     50249
Name: host_is_superhost, dtype: int64

In [96]:
#check data type is correct

df['host_is_superhost'].dtype

dtype('int64')

In [97]:
#change from float to int host_total_listings_count

df['host_total_listings_count'] = df['host_total_listings_count'].astype(int)

In [98]:
#verification step

df['host_total_listings_count'].dtype

dtype('int32')

In [99]:
# transform host_has_profile_pic into int by replacing 't' with 1 and 'f' with 0 

df['host_has_profile_pic'] = df['host_has_profile_pic'].replace('f',0)
df['host_has_profile_pic'] = df['host_has_profile_pic'].replace('t',1)

In [100]:
#verification step 

df['host_has_profile_pic'].value_counts()

1    278696
0       903
Name: host_has_profile_pic, dtype: int64

In [101]:
#check data type 

df['host_has_profile_pic'].dtype

dtype('int64')

In [102]:
#we repeat the same steps for host_identity_verified 

df['host_identity_verified'] = df['host_identity_verified'].replace('f',0)
df['host_identity_verified'] = df['host_identity_verified'].replace('t',1)

In [103]:
#verification step

df['host_identity_verified'].value_counts()

1    201263
0     78336
Name: host_identity_verified, dtype: int64

In [104]:
#check data type is correct

df['host_identity_verified'].dtype

dtype('int64')

In [105]:
#repeat steps for instant_bookable

df['instant_bookable'] = df['instant_bookable'].replace('f',0)
df['instant_bookable'] = df['instant_bookable'].replace('t',1)

In [106]:
#verification step

df['instant_bookable'].value_counts()

0    163992
1    115607
Name: instant_bookable, dtype: int64

In [107]:
#check data type is correct

df['instant_bookable'].dtype

dtype('int64')

The wrangling process is not completed 100% due to 2 remaining task: 

1. amenities column contains as cell values lists with the amenities, which I need to transform
2. price column contains values in local currency, which I need to standardize by transforming it to only 1 currency (USD)

Until I find the solution for these 2 steps, I will export the work completed so far into a new csv file to save all the changes made and not re-run all commands until now with every start. 

In [126]:
#send df to csv

df.to_csv(r'C:\\Users\Domnica Maria\Documents\1. PROJECTS\Maven-Vacation-Challenge\Airbnb-dataset.csv')

... **to be continued** 

see Jupyter Notebook 3